### ISSUE TITLE
Incorrect Return Value Compared to Geth

### ISSUE THREAD

**Issue by d-xo** [ethereum/hevm#289]
Another one discovered via differential fuzzing:

`cabal run test -- --quickcheck-replay=783978 -p "random-contract-concrete-call". reproduced at `dbe984a96042a3873ecdffb84c60151d6229689a`

**Comment by msooseth** [comment]
Notes in this comment.

Contract:
```
a = Tracing.OpContract [OpPush (Lit 0x9),OpPush (Lit 0x3),OpPush (Lit 0x5),OpPush (Lit 0x9),OpPush (Lit 0x1),OpPush (Lit 0x2),OpPush (Lit 0xa),OpPush (Lit 0x2),OpPush (Lit 0x2),OpPush (Lit 0x5),OpPush (Lit 0x0),OpShr,OpPush (Lit 0x9),OpPush (Lit 0x2),OpPush (Lit 0x6),OpPush (Lit 0x6),OpPush (Lit 0x5),OpPush (Lit 0x5),OpPush (Lit 0x1),OpPush (Lit 0x9),OpPush (Lit 0x3),OpCalldataload,OpPush (Lit 0x7),OpPush (Lit 0x5),OpPush (Lit 0x0),OpPush (Lit 0x8),OpPush (Lit 0x2),OpPush (Lit 0x7),OpCalldataload,OpPush (Lit 0x3),OpPush (Lit 0x5),OpDup 8,OpPush (Lit 0xa),OpMstore,OpPush (Lit 0x0),OpPush (Lit 0x3),OpPush (Lit 0x0),OpPush (Lit 0x8),OpPush (Lit 0x6),OpGt,OpPush (Lit 0x5),OpPush (Lit 0x9),OpPush (Lit 0x3),OpPush (Lit 0x3),OpPush (Lit 0x4),OpPush (Lit 0x5),OpPush (Lit 0xa),OpPush (Lit 0x5),OpPush (Lit 0x1),OpPush (Lit 0x7),OpPush (Lit 0x4),OpCalldataload,OpPush (Lit 0x0),OpCalldataload,OpPush (Lit 0x9),OpPush (Lit 0x3),OpCoinbase,OpPush (Lit 0x1),OpSub,OpPush (Lit 0x9),OpCalldatacopy,OpPush (Lit 0x8),OpPush (Lit 0x8),OpPush (Lit 0xa),OpPush (Lit 0x5),OpPush (Lit 0x5),OpPush (Lit 0x6),OpPush (Lit 0xa),OpPush (Lit 0x1),OpCalldataload,OpPush (Lit 0x1),OpPush (Lit 0x7),OpPush (Lit 0x40),OpPush (Lit 0x0),OpReturn]
```

Differing values:

```
return value computed      : Just "\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\a\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL"
```
and
```
evmtool's return value     : 0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000700000000000000000000000000000000000000000000
```

Notice the `7` vs the `a`. One has more bits set: `a` can only be represented by `101` (8+2) sequence but `7` can only be represented by `111` (4+2+1). So it's likely a real bug.

**Comment by d-xo** [comment]
Can we see the opcode where the traces differ?

**Comment by msooseth** [comment]
Good idea, actually, I'll try! The fuzzer found another simplification fuzz bug, I wanna get that fixed first so we can work more smoothly -- it may trigger during our PRs. So my queue got preemted :laughing: I'll get back to this in an  hour or so!

**Comment by msooseth** [comment]
Got it. Fix in #316 Problem was wraparound in `CallDataCopy`. `OpSub` did the trick here, with two small numbers being subtracted.
